src/lib.rs

```diff
@@ -0,0 +1,177 @@
+    deprecated,
+    unconditional_recursion,
+    unknown_lints,
+    unreachable_code,
+    unused_mut
+)]
+
+#[macro_use]
+extern crate log;
+#[macro_use]
+extern crate serde_derive;
+use elasticsearch::http::transport::{SingleNodeConnectionPool, TransportBuilder};
+use elasticsearch::Elasticsearch;
+use std::collections::{BTreeMap, HashMap};
+use std::time::Duration;
+
+/// Generic collector of Elasticsearch metrics
+pub mod collection;
+/// Metric
+pub mod metric;
+mod options;
+pub use options::ExporterOptions;
+
+mod metadata;
+
+pub(crate) mod metrics;
+
+const NAMESPACE: &'static str = "elasticsearch";
+
+/// The default [`Histogram`] buckets for Elasticsearch.
+pub const DEFAULT_BUCKETS: &[f64; 19] = &[
+    0.020, 0.040, 0.060, 0.080, 0.1, // <= 100ms
+    0.250, 0.500, 0.750, 1.0, // <= 1 second
+    2.0, 4.0, 6.0, 8.0, 10.0, // <= 10 seconds
+    20.0, 30.0, 40.0, 50.0, 60.0, // <= 1 minute
+];
+
+/// Labels type with ordered keys
+pub type Labels = BTreeMap<String, String>;
+
+/// Collection labels
+pub type CollectionLabels = BTreeMap<String, Vec<String>>;
+
+/// Exporter polling intervals
+pub type ExporterPollIntervals = BTreeMap<String, Duration>;
+
+/// Elasticsearch exporter
+#[derive(Debug, Clone)]
+pub struct Exporter {
+    /// Elasticsearch client instance
+    pub client: Elasticsearch,
+    /// Exporter options
+    pub options: ExporterOptions,
+    /// Constant exporter labels, e.g.: cluster
+    pub const_labels: HashMap<String, String>,
+
+    /// Node ID to node name map for adding extra metadata labels
+    /// {"U-WnGaTpRxucgde3miiDWw": "m1-supernode.example.com"}
+    pub id_to_name: HashMap<String, String>,
+}
+
+impl Exporter {
+    /// Spawn exporter
+    pub async fn new(options: ExporterOptions) -> Result<Self, Box<dyn std::error::Error>> {
+        let connection_pool = SingleNodeConnectionPool::new(options.elasticsearch_url.clone());
+        let transport = TransportBuilder::new(connection_pool)
+            .timeout(options.elasticsearch_global_timeout)
+            .build()?;
+
+        let client = Elasticsearch::new(transport);
+        info!("Elasticsearch::ping");
+        let _ = client.ping().send().await?;
+
+        let id_to_name = metadata::nodes_id_map(&client).await?;
+        let cluster_name = metadata::cluster_name(&client).await?;
+
+        let mut const_labels = HashMap::new();
+        let _ = const_labels.insert("cluster".into(), cluster_name);
+
+        Ok(Self {
+            client,
+            options,
+            const_labels,
+            id_to_name,
+        })
+    }
+
+    /// Spawn collectors
+    pub async fn spawn(self) {
+        info!("Spawned");
+        Self::spawn_cat(self.clone());
+        Self::spawn_cluster(self.clone());
+        Self::spawn_nodes(self.clone());
+    }
+
+    #[allow(unused)]
+    fn spawn_cluster(exporter: Self) {
+        let _ = tokio::spawn(metrics::_cluster::health::poll(exporter.clone()));
+    }
+
+    #[allow(unused)]
```

Comment:
Are these `#[allow(unused)]` attributes intended?

Reply:
yes for convenience, when I did not have metrics switch configuration, now those are redundant